doc/developer/design/20240625_subsource_deprecation_blue_green_sources.md

```diff
@@ -0,0 +1,318 @@
+### Implementation Plan
+
+1. Separate the planning of *sources* and *subsources* such that sources are fully
+   planned before any linked subsources. Part of this work has been prototyped in:
+   https://github.com/MaterializeInc/materialize/pull/27320
+
+2. Introduce the concept of *read-only tables* that link to a source ingestion
+   and an upstream 'reference' specifying which upstream table to ingest into this table.
+
+3. Update the CREATE TABLE statement to allow creation of a read-only table
+   with a reference to a source and an upstream-reference:
+   (`CREATE TABLE <new_table> FROM SOURCE <source> (upstream_reference)`)
+
+4. Update planning for `CREATE TABLE` to include a purification step akin to the
+   purification for `ALTER SOURCE .. ADD SUBSOURCE`. This will verify the upstream
+   permissions, schema, etc for the newly added source-fed table.
+
+5. Update the underlying *source* (ingestion) and source_export/ingestion_export
+   structures to allow more than one *source_export* to refer to an upstream
+   reference, and each *source_export* to have its own projection of the upstream data
+
+5. Update the storage controller to use both *subsources* and *read-only tables*
+   as *source_exports* for existing multi-output sources (postgres & mysql)
+
+6. Update the source rendering operators to handle the new structures and allow
+   outputting the same upstream table to more than one souce export.
+
+7. Migrate existing sources to the new source model (make all sources 'multi-output' sources):
+   - For existing sources where the primary relation has the data (e.g. Kafka sources):
+     - Create a read-only table for the current primary relation called `<source>`
+     - Rename the source itself `<source>_progress` and drop the existing
+      `<source>_progress` object
+   - For existing multi-output sources (e.g. Postgres & MySQL sources):
+     - Convert subsources to read-only tables with the same name
+     - Rename the source itself `<source>_progress` and drop the existing
+      `<source>_progress` object
+
+8. Remove subsource purification logic from `purify_create_source`, subsource statement parsing,
+   and related planning code
+
+
+### Core Implementation Details
+
+#### SQL Parsing & Planning
+
+Our existing `CREATE SUBSOURCE` statements look like this:
+```sql
+CREATE SUBSOURCE <name> (<cols>) OF SOURCE <source_name> WITH (EXTERNAL REFERENCE = <upstream table name>)
+```
+though these are mostly hidden from users, they are available if a user runs `SHOW CREATE SOURCE <existing_subsource>`
+
+Our existing `CREATE TABLE` statement looks like:
+```sql
+CREATE TABLE <name> (<cols>) WITH (<options>)
+```
+
+We would update the `CREATE TABLE` statement to look similar to `CREATE SUBSOURCE` and
+be able to optionally reference an upstream source and reference using `OF SOURCE` and `EXTERNAL REFERENCE`:
+```sql
+CREATE TABLE <name> (<cols>) OF SOURCE <source_name> WITH (EXTERNAL REFERENCE = <upstream table name>)
+```
```

Comment:
I think it's worth brainstorming some syntax alternatives and seeing what we like best. A few ideas to get things started:

```sql
CREATE TABLE <name> (<cols>)
FROM SOURCE <src_name> ([<upstream_schema>.]<upstream_name>);

CREATE EXTERNAL TABLE <name> (<cols>)
SOURCE <src_name> ([<upstream_schema>.]<upstream_name>);

CREATE SOURCE TABLE <name> (<cols>)
SOURCE <src_name> ([<upstream_schema>.]<upstream_name>)

CREATE EXTERNAL TABLE <name> (<cols>)
WITH (
    SOURCE = src,
    EXTERNAL NAME = [<upstream_schema>.]<upstream_name>
);
```

On a conceptual level, I think we should consider whether we want to call these "foreign tables" (to match PostgreSQL), "external tables" (to match ... approximately everyone else), or "source tables" (to match our own terminology, at the risk of maybe being confusing?).

Reply:
> On a conceptual level, I think we should consider whether we want to call these "foreign tables" (to match PostgreSQL), "external tables" (to match ... approximately everyone else), or "source tables" (to match our own terminology, at the risk of maybe being confusing?).

Yes I think this is a good discussion to have!

If we think about the end-to-end flow for a user who wants to ingest data from an upstream postgres table, my current proposal looks like:

``` 
CREATE CONNECTION .. TO POSTGRES (HOST .., PORT .., ..) WITH (VALIDATE ..);

CREATE SOURCE .. FROM POSTGRES CONNECTION .. (PUBLICATION .., TEXT COLUMNS ..) WITH (RETAIN HISTORY FOR ..)

CREATE TABLE .. (col1, col2) OF SOURCE .. WITH (EXTERNAL REFERENCE = ..)
```

which seems a little awkward now that I look at it. I think I'm leaning towards the term `FROM SOURCE` rather than `OF SOURCE`, and would actually propose:
`CREATE TABLE .. (cols) FROM SOURCE (SOURCE = <source_name>, REFERENCE = <external_reference>)`
rather than including the `EXTERNAL REFERENCE` as a `WITH` option

Reply:
I think spelling that last option like so:

```
CREATE TABLE .. (cols) FROM SOURCE <source_name> (REFERENCE <external_reference>)
```

would be more in line with existing similar syntax, like `CREATE SOURCE .. FROM KAFKA CONNECTION <connection_name> (KEY = VALUE, ...)`.

Reply:
I didn't know that under the hood we used `OF SOURCE` to create subsources, but +1 on using `FROM` instead.

The thought of external tables also ran through my mind reading through the design doc, but since we actually ingest the upstream data into our own storage layer, this could create more confusion than it would be a helpful reference?

Reply:
The only point of contention for me here is the point I raised earlier about dropping the options that allow you to skip manually creating each table after creating a source. Recapping, I think we could default to not creating tables by default (and encourage users — especially dbt users — to go this route), but it'd be a regression in usability to remove the options completely.

Reply:
> The thought of external tables also ran through my mind reading through the design doc, but since we actually ingest the upstream data into our own storage layer, this could create more confusion than it would be a helpful reference?

Yes I think this would introduce conceptual confusion, since you're not actually creating a view of the upstream table, you're actually copying the upstream table's data into materialize, so it's quite different than postgres' FDW for example

Reply:
FWIW +1 from me on `CREATE TABLE <name> FROM SOURCE <source> ...`

Totally agree with the points of `CREATE [EXTERNAL | FOREIGN] TABLE ...`. It seems like it could be a good fit here but because we copy data into Materialize while other databases only reference the external data, it's not an exact fit.

Only thought is do we need the `SOURCE` keyword in the statement? Could we instead make the syntax `CREATE TABLE <name> FROM <source>`?

Reply:
> Only thought is do we need the SOURCE keyword in the statement? Could we instead make the syntax CREATE TABLE <name> FROM <source>?

I think it makes sense to be explicit here, though we don't require the object qualifier in `CREATE SINK`. Debatable! There's only one object type allowed in this case, though, so it might be a helpful pointer to include `SOURCE`.